File: app/src/main/java/im/point/dotty/feed/FeedAdapter.java
package im.point.dotty.feed;

import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import im.point.dotty.R;
import im.point.dotty.model.Post;

public class FeedAdapter<T extends Post> extends RecyclerView.Adapter<PostHolder> {
    private List<T> list;

    FeedAdapter() {
        setHasStableIds(true);
        this.list = new ArrayList<>();
    }

    public List<T> getList() {
        return list;
    }

    public void setList(List<T> list) {
        this.list = list;
        notifyDataSetChanged();
    }

    @Override
    public PostHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_post, parent, false);
        return new PostHolder(view);
    }

    @Override
    public void onBindViewHolder(PostHolder holder, int position) {
        holder.bind(list.get(position));
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    @Override
    public long getItemId(int position) {
        return list.get(position).getPageId();
    }
}

File: app/src/main/java/im/point/dotty/feed/FeedAdapter.kt
/*
 * Copyright (c) 2019-2021 Pavel A. Sokolov
 */
package im.point.dotty.feed

import android.graphics.Bitmap
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import im.point.dotty.R
import im.point.dotty.model.CompletePost
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emptyFlow

class FeedAdapter<T : CompletePost<*>> internal constructor(val scope: CoroutineScope) : RecyclerView.Adapter<PostHolder<T>>() {
    var list: List<T> = listOf()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    var avatarProvider: (name: String) -> Flow<Bitmap> = { emptyFlow() }
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    var imagesProvider: (name: String) -> Flow<List<Bitmap>> = { emptyFlow() }
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    var onPostClicked: (item: T) -> Unit = {}
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    var onUserClicked: (id: Long, login: String) -> Unit = { _, _ -> }
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    var onTagClicked: (tag: String) -> Unit = {}
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PostHolder<T> =
            LayoutInflater.from(parent.context).inflate(R.layout.post, parent, false)
                    .let { PostHolder(it, scope) }

    override fun onBindViewHolder(holder: PostHolder<T>, position: Int) = with(list[position]) {
        holder.bind(this, avatarProvider(authorLogin), imagesProvider(id),
                onPostClicked, onUserClicked, onTagClicked)
    }

    override fun getItemCount(): Int = list.size

    override fun getItemId(position: Int): Long = list[position].id.hashCode().toLong()

    init {
        setHasStableIds(true)
    }
}